lib/galaxy/objectstore/_caching_base.py

```diff
@@ -0,0 +1,384 @@
+    ObjectNotFound,
+)
+from galaxy.objectstore import ConcreteObjectStore
+from galaxy.util import (
+    directory_hash_id,
+    ExecutionTimer,
+    unlink,
+)
+from galaxy.util.path import safe_relpath
+from ._util import fix_permissions
+from .caching import (
+    CacheTarget,
+    InProcessCacheMonitor,
+)
+
+log = logging.getLogger(__name__)
+
+
+class CachingConcreteObjectStore(ConcreteObjectStore):
+    staging_path: str
+    extra_dirs: Dict[str, str]
+    config: Any
+    cache_updated_data: bool
+    enable_cache_monitor: bool
+    cache_size: int
+    cache_monitor: Optional[InProcessCacheMonitor] = None
+    cache_monitor_interval: int
+
+    def _construct_path(
+        self,
+        obj,
+        base_dir=None,
+        dir_only=None,
+        extra_dir=None,
+        extra_dir_at_root=False,
+        alt_name=None,
+        obj_dir=False,
+        in_cache=False,
+        **kwargs,
+    ):
+        # extra_dir should never be constructed from provided data but just
+        # make sure there are no shenannigans afoot
+        if extra_dir and extra_dir != os.path.normpath(extra_dir):
+            log.warning("extra_dir is not normalized: %s", extra_dir)
+            raise ObjectInvalid("The requested object is invalid")
+        # ensure that any parent directory references in alt_name would not
+        # result in a path not contained in the directory path constructed here
+        if alt_name:
+            if not safe_relpath(alt_name):
+                log.warning("alt_name would locate path outside dir: %s", alt_name)
+                raise ObjectInvalid("The requested object is invalid")
+            # alt_name can contain parent directory references, but S3 will not
+            # follow them, so if they are valid we normalize them out
+            alt_name = os.path.normpath(alt_name)
+
+        object_id = self._get_object_id(obj)
+        rel_path = os.path.join(*directory_hash_id(object_id))
+
+        if extra_dir is not None:
+            if extra_dir_at_root:
+                rel_path = os.path.join(extra_dir, rel_path)
+            else:
+                rel_path = os.path.join(rel_path, extra_dir)
+
+        # for JOB_WORK directory
+        if obj_dir:
+            rel_path = os.path.join(rel_path, str(object_id))
+        if base_dir:
+            base = self.extra_dirs.get(base_dir)
+            assert base
+            return os.path.join(base, rel_path)
+
+        if not dir_only:
+            rel_path = os.path.join(rel_path, alt_name if alt_name else f"dataset_{object_id}.dat")
+
+        if in_cache:
+            return self._get_cache_path(rel_path)
+
+        return rel_path
+
+    def _get_cache_path(self, rel_path: str) -> str:
+        return os.path.abspath(os.path.join(self.staging_path, rel_path))
+
+    def _in_cache(self, rel_path: str) -> bool:
+        """Check if the given dataset is in the local cache and return True if so."""
+        cache_path = self._get_cache_path(rel_path)
```

Comment:
There are open bugs (https://github.com/galaxyproject/galaxy/issues/15283, maybe https://github.com/galaxyproject/galaxy/issues/16438), where the cached copy is of size 0 while the remote file is complete. Should we check the size ?

Reply:
Comment in ``_get_filename``:

```python
        # Check if the file exists in the cache first, always pull if file size in cache is zero
        if self._in_cache(rel_path) and (dir_only or os.path.getsize(self._get_cache_path(rel_path)) > 0):
            return cache_path
```

I feel like I vaguely prefer doing it a layer above ``_in_cache`` like that. And it seems like it has the behavior you're looking for as long as ``sync_cache`` is unset of passed as True.

So I am anxious this is not the right approach and results in more API calls. For https://github.com/galaxyproject/galaxy/issues/15283 - hmm... maybe we make:

```
        file_ok = self._download(rel_path)
        fix_permissions(self.config, self._get_cache_path(rel_path_dir))
        return file_ok
```

in _pull_into_cache into:

```
        file_ok = self._download(rel_path)
        if file_ok:
             fix_permissions(self.config, self._get_cache_path(rel_path_dir))
        else:
              unlink(self._get_cache_path(rel_path), ignore_errors=True)
        return file_ok
```

This would seem to prevent an important class of errors. I could see it fixing https://github.com/galaxyproject/galaxy/issues/15283  but I doubt it is the problem with https://github.com/galaxyproject/galaxy/issues/16438. I think that problem would be better address by bailing on object store creation if a target job cache directory is not writable? I know this isn't a perfect fix for https://github.com/galaxyproject/galaxy/issues/16438 but it wouldn't leave us in an inconsistent state where we're having to deduce why things didn't work and wonder what exception handler is broken.

Do you have any objections to these changes as an effort to address those two issues?

Reply:
No objections, sounds like a solid improvement, thank you!

Reply:
And look we will be fixing 6 object stores at once instead of making changes to copy and pasted code shared between all of them but only fixed in one 😆